### ISSUE TITLE
Rule to disallow closing over modified variables

### ISSUE THREAD

**Issue by RyanCavanaugh** [palantir/tslint#81]
This is a common source of bugs; the straightforward cases are pretty easy to define:

``` javascript
// Error: Closure over variable declared in 'for' loop
for(var i = 0; i < 10; i++) {
    window.setTimeout(() => console.log(i), 50);
}

// Error: Closure over variable used in assignment position in same loop construct body
var i = 0;
while(i < 100) {
    i++;
    window.setTimeout(() => console.log(i), 50);    
}

// Error: Closure over variable used in assignment position after closure creation
var i = 0;
window.setTimeout(() => console.log(i), 50);
i++;
```

**Comment by awerlang** [comment]
+1

**Comment by bafolts** [comment]
Is there a rule to disallow closures altogether, namely, using any variable in a function that isn't an argument, this, super, or a variable you created in that function? Besides being a common source of bugs it is also a common source of memory leaks. If I read this description right, is the goal to not allow closures? As I want to implement this feature I would submit a PR.

**Comment by JoshuaKGoldberg** [comment]
This rule seems less relevant now that `var` is so strongly discouraged. It's also been 5 years without an implementation. Feel free to comment here if you think this rule should still be implemented!
